src/transformers/models/jetmoe/modeling_jetmoe.py

```diff
@@ -0,0 +1,1708 @@
+    def __init__(self, config: JetMoeConfig):
+        super(JetMoeMoA, self).__init__()
+
+        self.num_experts = config.num_local_experts
+        self.input_size = config.hidden_size
+        self.hidden_size = config.kv_channels * config.num_key_value_heads
+        self.top_k = config.num_experts_per_tok
+        self.bias = torch.nn.Parameter(torch.empty(self.input_size))
+
+        self.input_linear = JetMoeParallelExperts(self.num_experts, self.input_size, self.hidden_size)
+        self.output_linear = JetMoeParallelExperts(self.num_experts, self.hidden_size, self.input_size)
+
+        self.router = JetMoeTopKGating(
+            input_size=self.input_size,
+            num_experts=self.num_experts,
+            top_k=self.top_k,
+        )
+
+    def map(self, layer_input):
+        """
+        Map input through the mixture of experts layer.
+
+        Args:
+            x (Tensor):
+                Input tensor.
+
+        Returns:
+            Tensor:
+                Output tensor.
+            Tensor:
+                Router logits.
+        """
+        bsz, length, emb_size = layer_input.size()
+        layer_input = layer_input.reshape(-1, emb_size)
+        index_sorted_experts, batch_index, batch_gates, expert_size, router_logits = self.router(layer_input)
+        self.topo_info = (index_sorted_experts, batch_index, batch_gates, expert_size)
+
+        expert_inputs = layer_input[batch_index]
+        expert_outputs = self.input_linear(expert_inputs, expert_size)
+
+        zeros = torch.zeros(
+            (bsz * length * self.top_k, self.hidden_size), dtype=expert_outputs.dtype, device=expert_outputs.device
+        )
+        layer_output = zeros.index_add(0, index_sorted_experts, expert_outputs)
+        layer_output = layer_output.view(bsz, length, self.top_k, -1)
+        return layer_output, router_logits
+
+    def reduce(self, layer_input):
+        """
+        Reduce the mapped output.
+
+        Args:
+            x (Tensor):
+                Mapped output tensor.
+
+        Returns:
+            Tensor:
+                Reduced output tensor.
+        """
+
+        bsz, length, k, emb_size = layer_input.size()
+        layer_input = layer_input.reshape(-1, emb_size)
+
+        index_sorted_experts, batch_index, batch_gates, expert_size = self.topo_info
+
+        expert_inputs = layer_input[index_sorted_experts]
+        expert_outputs = self.output_linear(expert_inputs, expert_size)
+
+        expert_outputs = expert_outputs * batch_gates[:, None]
+
+        zeros = torch.zeros((bsz * length, self.input_size), dtype=expert_outputs.dtype, device=expert_outputs.device)
+        layer_output = zeros.index_add(0, batch_index, expert_outputs)
+        layer_output = layer_output.view(bsz, length, self.input_size)
+        layer_output = layer_output + self.bias
+        return layer_output
+
+    def forward(self, layer_input):
+        raise NotImplementedError("This module is not meant to be used directly.")
```

Comment:
it's not the entire modeul that is not meant to be used directly, just not mean to support call and forward.